README
Net-Google-SafeBrowsing2 version 1.11
=====================================

Net::Google::SafeBrowsing2 implements the Google Safe Browsing v2 API.
Net::Google::SafeBrowsing2::Lookup implements the Google Safe Browsing v2 Lookup API (up to 10,000 URL checks a day).

The library passes most of the unit tests listed in the API documentation. 
See the documentation (http://code.google.com/apis/safebrowsing/developers_guide_v2.html) for more details about the failed tests.

The Google Safe Browsing database must be stored and managed locally.
Net::Google::SafeBrowsing2::Sqlite uses Sqlite as the storage back-end, Net::Google::SafeBrowsing2::MySQL uses MySQL.
Other storage mechanisms (databases, memory, etc.) can be added and used transparently with this module.

You may want to look at "Google Safe Browsing v2: Implementation Notes" (http://www.zscaler.com/research/Google%20Safe%20Browsing%20v2%20API.pdf), a collection of notes and real-world numbers about the API. This is intended for people who want to learn more about the API, whether as a user or to make their own implementation.

The source code is available on github at https://github.com/juliensobrier/Net-Google-SafeBrowsing2.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  LWP::UserAgent
  URI
  Digest::SHA
  List::Util
  constant
  Test::More
  Text::Trim
  Digest::HMAC_SHA1
  MIME::Base64::URLSafe
  MIME::Base64
  String::HexConvert
  File::Slurp
  IO::Socket::SSL


COPYRIGHT AND LICENCE

Copyright (C) 2013 by Julien Sobrier

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.



Changes
Revision history for Perl extension Net::Google::SafeBrowsing2.

0.1  Wed Jul 21 16:35:48 2010
	- original version

0.2 Fri Nov 05 14:45:48 2010
        - add support for MAC

0.5 Mon Mar 07 16:55:00 2011
		- add MySQl back-end storage
        - introduce base class Net::Google::SafeBrowsing2::DBI
		- update documentation

    Tue Apr 05 14:40:00 2011
		- fix Net::Google::SafeBrowsing2::DBI to work with all database types

0.6 Mon Jul 25 13:21:00 2011
		- Handle local database reset

0.7 Mon Jul 25 15:18:00 2011
		- Add local_lookup to perform a lookup against the local database only for debugging purpose.
		- Fix duplicate insert of add chunks and sub chunks.

0.8 Tue Aug 02 13:05:00 2011
		- Fix chunk retrieval with MySQL backend (VARCHAR to VARBINARY)
		- Reduce the number of full hashes requested 

0.9 Tue Aug 09 15:26:00 2011
		- Fix bug with local whitelisting (sub chunks).
		- Fix the parsing of full hashes.

1.0 Wed Aug 10 15:11:00
		- Storage: add keep_all argument to keep expired full hashes in the database
		- Separate the error output from debug output
		- A lot of testing

1.01 Wed Aug 26 13:49:00
		- Use String::HexConvert for faster hex_to_ascii

1.02 Thu Nov 10 10:00:00
		- Fix uninitialized $self->{errors} variable

1.03 Thu Nov 21 17:24:00
		- no code change
		- source code available on github at https://github.com/juliensobrier/Net-Google-SafeBrowsing2
		- minor documentation updates

1.04 Wed Dec 07 10:26:00
		- add Net::Google::SafeBrowsing2::Lookup (http://code.google.com/apis/safebrowsing/lookup_guide.html)

	 Wed Dec 07 10:49:00
		- Net::Google::SafeBrowsing2::Sqlite, index s_chunks_unique was created at the wrong place

	Tue Dec 12 15:41:00
		- Net::Google::SafeBrowsing2::Lookup 0.2: documentation update

1.05 Thu Jan 12 14:26:00
		- No code change. Move C<local_lookup> to PRIVATE FUNCTIONS to avoid confusions.

1.05 Mon May 07 10:22:00
		- Added Postgres storage from njohnston

1.06 Sun Jun 03 21:40:00
		- Update DBI and MySQL back-end to keep empty sub chunks. Fix index for s_chunks table
		- MySQL: truncate prefixes longer than 8 bytes
		- Add a warning about the number of updates required with an empty database
		- Storage: reset function is required

1.07 Fri Jun 08 18:38:00
		- Import/Export feature for add chunks and sub chunks. Can be used to move from one back-end storage to another.

1.08 Mon Apr 08 2013 07:48:00
		- Fix Po documentation (contribution)

1.09 Wed Jun 05 2013 20:29:00
		- Use HTTPS to access safebrowsing.clients.google.com/.

1.10 Tue Dec 31 2013 22:15:00
		- Force IPv4 to solve bug on CentOS.

1.11 Fri Jul 11 2014 21:41:00
		- Add dependency on IO::Socket::SSL
		- Remove dependency on Net::IPAddress

lib/Net/Google/SafeBrowsing2/MySQL.pm
package Net::Google::SafeBrowsing2::MySQL;

use strict;
use warnings;

use base 'Net::Google::SafeBrowsing2::DBI';

use Carp;
use DBI;
use List::Util qw(first);


our $VERSION = '1.0';


=head1 NAME

Net::Google::SafeBrowsing2::MySQL - MySQL as back-end storage for the Google Safe Browsing v2 database

=head1 SYNOPSIS

  use Net::Google::SafeBrowsing2::MySQL;

  my $storage = Net::Google::SafeBrowsing2::MySQL->new(host => '127.0.0.1', database => 'GoogleSafeBrowsingv2');
  ...
  $storage->close();

=head1 DESCRIPTION

This is an implementation of L<Net::Google::SafeBrowsing2::Storage> using MySQL.

=cut


=head1 CONSTRUCTOR

=over 4

=head2 new()

Create a Net::Google::SafeBrowsing2::MySQL object

  my $storage = Net::Google::SafeBrowsing2::MySQL->new(
      host     => '127.0.0.1', 
      database => 'GoogleSafeBrowsingv2', 
      username => 'foo', 
      password => 'bar'
  );

Arguments

=over 4

=item host

Required. MySQL host name

=item database

Required. MySQL database name to connect to.

=item username

Required. MySQL username.

=item password

Required. MySQL password.

=item port

Optional. MySQL port number to connect to.

=item keep_all

Optional. Set to 1 to keep old information (such as expiring full hashes) in the database. 0 (delete) by default.

=back


=back

=cut

sub new {
	my ($class, %args) = @_;

	my $self = { # default arguments
		host		=> '127.0.0.1',
		database	=> 'GoogleSafeBrowsingv2',
		port		=> 3306,
		keep_all	=> 0,

		%args,
	};

	bless $self, $class or croak "Can't bless $class: $!";


	$self->init();

    return $self;
}

=head1 PUBLIC FUNCTIONS

=over 4

See L<Net::Google::SafeBrowsing2::Storage> for a complete list of public functions.

=head2 close()

Cleanup old full hashes, and close the connection to the database.

  $storage->close();


=back

=cut

sub init {
	my ($self, %args) = @_;

	$self->{dbh} = DBI->connect("DBI:mysql:database=" . $self->{database} . ";host=" . $self->{host} . ";port=" . $self->{port}, $self->{username}, $self->{password}, {'RaiseError' => 1});

	my @tables = $self->{dbh}->tables;

	if (! defined first { $_ =~ '`updates`' } @tables) {
		$self->create_table_updates();
	}
	if (! defined first { $_ =~ '`a_chunks`' } @tables) {
		$self->create_table_a_chunks();
	}
	if (! defined first { $_ =~ '`s_chunks`' } @tables) { 
		$self->create_table_s_chunks();
	}
	if (! defined first { $_ =~ '`full_hashes`' } @tables) {
		$self->create_table_full_hashes();
	}
	if (! defined first { $_ =~ '`full_hashes_errors`' } @tables) { 
		$self->create_table_full_hashes_errors();
	}
	if (! defined first { $_ =~ '`mac_keys`' } @tables) { 
		$self->create_table_mac_keys();
	}

	# Remove indexes previously created
	$self->{dbh}->do("ALTER TABLE `a_chunks` DROP INDEX `a_chunks_hostkey`") if ($self->{dbh}->do("SHOW INDEX FROM `a_chunks`WHERE KEY_NAME = 'a_chunks_hostkey'") > 0);
	$self->{dbh}->do("ALTER TABLE `a_chunks` DROP INDEX `a_chunks_num_list`") if ($self->{dbh}->do("SHOW INDEX FROM `a_chunks`WHERE KEY_NAME = 'a_chunks_num_list'") > 0);

	$self->{dbh}->do("ALTER TABLE `s_chunks` DROP INDEX `s_chunks_hostkey`") if ($self->{dbh}->do("SHOW INDEX FROM `s_chunks`WHERE KEY_NAME = 's_chunks_hostkey'") > 0);
	$self->{dbh}->do("ALTER TABLE `s_chunks` DROP INDEX `s_chunks_num`") if ($self->{dbh}->do("SHOW INDEX FROM `s_chunks`WHERE KEY_NAME = 's_chunks_num'") > 0);
	$self->{dbh}->do("ALTER TABLE `s_chunks` DROP INDEX `s_chunks_num_list`") if ($self->{dbh}->do("SHOW INDEX FROM `s_chunks`WHERE KEY_NAME = 's_chunks_num_list'") > 0);
}


sub create_table_updates {
	my ($self, %args) = @_;

	my $schema = qq{	
		CREATE TABLE updates (
			last INT NOT NULL DEFAULT '0',
			wait INT NOT NULL DEFAULT '0',
			errors INT NOT NULL DEFAULT '1800',
			list VARCHAR( 50 ) NOT NULL
		);
	}; # Need to handle errors

	$self->{dbh}->do($schema);
}

sub create_table_a_chunks {
	my ($self, %args) = @_;

	my $schema = qq{
		CREATE TABLE a_chunks (
			hostkey VARBINARY( 8 ),
			prefix VARBINARY( 8 ),
			num INT NOT NULL,
			list VARCHAR( 50 ) NOT NULL
		);
	};

	$self->{dbh}->do($schema);

	my $index = qq{
		CREATE INDEX a_chunks_hostkey ON a_chunks (
			hostkey
		);
	};
# 	$self->{dbh}->do($index);

	$index = qq{
		CREATE INDEX a_chunks_num_list ON a_chunks (
			num,
			list
		);
	};
# 	$self->{dbh}->do($index);

	$index = qq{
		CREATE UNIQUE INDEX a_chunks_unique ON a_chunks (
			hostkey,
			prefix,
			num,
			list
		);
	};
	$self->{dbh}->do($index);
}

sub create_table_s_chunks {
	my ($self, %args) = @_;

	my $schema = qq{
		CREATE TABLE s_chunks (
			hostkey VARBINARY( 8 ),
			prefix VARBINARY( 8 ),
			num INT NOT NULL,
			add_num INT DEFAULT 0,
			list VARCHAR( 50 ) NOT NULL
		);
	};

	$self->{dbh}->do($schema);

	my $index = qq{
		CREATE INDEX s_chunks_hostkey ON s_chunks (
			hostkey
		);
	};
# 	$self->{dbh}->do($index);

	$index = qq{
		CREATE INDEX s_chunks_num ON s_chunks (
			num
		);
	};
# 	$self->{dbh}->do($index);

	$index = qq{
		CREATE INDEX s_chunks_num_list ON s_chunks (
			num,
			list
		);
	};
# 	$self->{dbh}->do($index);

	$index = qq{
		CREATE UNIQUE INDEX s_chunks_unique ON s_chunks (
			hostkey,
			prefix,
			num,
			add_num,
			list
		);
	};
	$self->{dbh}->do($index);
}

sub create_table_full_hashes {
	my ($self, %args) = @_;

	my $schema = qq{
		CREATE TABLE full_hashes (
			id INT AUTO_INCREMENT PRIMARY KEY,
			num INT,
			hash VARBINARY( 32 ),
			list VARCHAR( 50 ),
			timestamp INT Default '0'
		);
	};

	$self->{dbh}->do($schema);

	my $index = qq{
		CREATE UNIQUE INDEX hash ON full_hashes (
			num,
			hash,
			list
		);
	};
	$self->{dbh}->do($index);
}

sub create_table_full_hashes_errors {
	my ($self, %args) = @_;

	my $schema = qq{
		CREATE TABLE full_hashes_errors (
			id INT AUTO_INCREMENT PRIMARY KEY,
			errors INT Default '0',
			prefix VARBINARY( 8 ),
			timestamp INT Default '0'
		);
	};

	$self->{dbh}->do($schema);
}

sub create_table_mac_keys{
	my ($self, %args) = @_;

	my $schema = qq{
		CREATE TABLE mac_keys (
			client_key VARCHAR( 50 ) Default '',
			wrapped_key VARCHAR( 50 ) Default ''
		);
	};

	$self->{dbh}->do($schema);
}


sub add_chunks_s {
	my ($self, %args) 	= @_;
	my $chunknum		= $args{chunknum}	|| 0;
	my $chunks			= $args{chunks}		|| [];
	my $list			= $args{'list'}		|| '';

	my $add = $self->{dbh}->prepare('INSERT IGNORE INTO s_chunks (hostkey, prefix, num, add_num, list) VALUES (?, ?, ?, ?, ?)');
	$self->{dbh}->{AutoCommit} = 0;

	foreach my $chunk (@$chunks) {
		$add->execute( $chunk->{host}, $chunk->{prefix}, $chunknum, $chunk->{add_chunknum}, $list );
	}

	if (scalar @$chunks == 0) { # keep empty chunks
		$add->execute( '', '',  $chunknum, '', $list );
	}

	$self->{dbh}->commit;
	$self->{dbh}->{AutoCommit} = 1;
}

sub add_chunks_a {
	my ($self, %args) 	= @_;
	my $chunknum		= $args{chunknum}	|| 0;
	my $chunks			= $args{chunks}		|| [];
	my $list			= $args{'list'}		|| '';

	my $add = $self->{dbh}->prepare('INSERT IGNORE INTO a_chunks (hostkey, prefix, num, list) VALUES (?, ?, ?, ?)');
	$self->{dbh}->{AutoCommit} = 0;

	foreach my $chunk (@$chunks) {
		# 32-byte prefix seen at chunk 69961
		# If this becomes more of a problem, the schema will have to be adjusted.
		if (length($chunk->{prefix}) > 8) {
			$chunk->{prefix} = substr $chunk->{prefix}, 0, 4;
		}

		$add->execute( $chunk->{host}, $chunk->{prefix}, $chunknum, $list );
	}

	if (scalar @$chunks == 0) { # keep empty chunks
		$add->execute( '', '', $chunknum, $list );
	}

	$self->{dbh}->commit;
	$self->{dbh}->{AutoCommit} = 1;
}

=head1 CHANGELOG

=over 4

=item 0.6

Remove some indexes to speed up INSERTs.

=item 0.5

Keep empty sub chunks. Shorten prefixes greater than 8 bytes (workaround to keep schema tight)

=item 0.4

Add option keep_all to keep expired full hashes. Useful for debugging.

=item 0.3

Use more efficient add_chunk_a and add_chunk_s functions.
Change data type for prefixes from VARCHAR to VARBINARY.


=back

=head1 SEE ALSO

See L<Net::Google::SafeBrowsing2> for handling Google Safe Browsing v2.

See L<Net::Google::SafeBrowsing2::Storage> for the list of public functions.

See L<Net::Google::SafeBrowsing2::Sqlite> for a back-end using Sqlite.

Google Safe Browsing v2 API: L<http://code.google.com/apis/safebrowsing/developers_guide_v2.html>


=head1 AUTHOR

Julien Sobrier, E<lt>jsobrier@zscaler.comE<gt> or E<lt>julien@sobrier.netE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Julien Sobrier

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.


=cut

1;

lib/Net/Google/SafeBrowsing2/Postgres.pm
package Net::Google::SafeBrowsing2::Postgres;

use strict;
use warnings;

use base 'Net::Google::SafeBrowsing2::DBI';

use Carp;
use DBI;
use List::Util qw(first);

our $VERSION = '0.02';

=head1 NAME

Net::Google::SafeBrowsing2::Postgres - Postgres as back-end storage for the
Google Safe Browsing v2 database

=head1 SYNOPSIS

  use Net::Google::SafeBrowsing2::Postgres;

  my $storage = Net::Google::SafeBrowsing2::Postgres->new(
    host     => '127.0.0.1',
    database => 'google_safe_browsing',
    username => $username,
    password => $password,
  );
  
  # ...

  $storage->close();

=head1 DESCRIPTION

This is an implementation of L<Net::Google::SafeBrowsing2::Storage> using
Postgres.

=head1 CONSTRUCTOR

=over 4

=head2 new()

Create a Net::Google::SafeBrowsing2::Postgres object

  my $storage = Net::Google::SafeBrowsing2::Postgres->new(
      host     => '127.0.0.1', 
      database => 'google_safe_browsing', 
      username => $username,
      password => $password,
  );

Arguments

=over 4

=item host

Specifies Postgres host name. Defaults to 127.0.0.1.

=item database

Specifies Postgres database name. Defaults to "google_safe_browsing".

=item username

Specifies the username for the Postgres connection. Required.

=item password

Specifies the password for the Postgres connection. Required.

=item keep_all

Optional. Set to 1 to keep old information (such as expiring full hashes)
in the database. 0 (delete) by default.

=back

=back

=cut

sub new {
  my ($class, %args) = @_;

  # Default arguments
  my $self = { 
    host     => '127.0.0.1',
    database => 'google_safe_browsing',
    keep_all => 0,

    %args,
  };

  if (!$self->{username}) {
    croak "username required";
  }

  if (!$self->{password}) {
    croak "password required";
  }

  bless $self, $class;

  $self->init();

  return $self;
}

=head1 PUBLIC FUNCTIONS

=over 4

See L<Net::Google::SafeBrowsing2::Storage> for a complete list of public functions.

=head2 close()

Cleanup old full hashes, and close the connection to the database.

  $storage->close();

=back

=cut

sub init {
  my ($self, %args) = @_;

  $self->{dbh} = DBI->connect(
    "DBI:Pg:dbname=" . $self->{database} . ";host=" . $self->{host},
    $self->{username},
    $self->{password},
    {
      RaiseError => 1,
    },
  );

  my @tables = $self->{dbh}->tables;

  # Postgres reports normal tables (compared to internal ones) with a
  # prefix of "public."
  if (! defined first { /public\.updates/ } @tables) {
    $self->create_table_updates();
  }
  if (! defined first { /public\.a_chunks/ } @tables) {
    $self->create_table_a_chunks();
  }
  if (! defined first { /public\.s_chunks/ } @tables) { 
    $self->create_table_s_chunks();
  }
  if (! defined first { /public\.full_hashes/ } @tables) {
    $self->create_table_full_hashes();
  }
  if (! defined first { /public\.full_hashes_errors/ } @tables) { 
    $self->create_table_full_hashes_errors();
  }
  if (! defined first { /public\.mac_keys/ } @tables) { 
    $self->create_table_mac_keys();
  }
}

# Overridden because Postgres uses SERIAL instead of AUTO_INCREMENT.
sub create_table_full_hashes {
  my ($self, %args) = @_;

  my $schema = qq{
    CREATE TABLE full_hashes (
      id SERIAL PRIMARY KEY,
      num INT,
      hash VARCHAR( 32 ),
      list VARCHAR( 50 ),
      timestamp INT Default '0'
    );
  };

  $self->{dbh}->do($schema);

  my $index = qq{
    CREATE UNIQUE INDEX hash ON full_hashes (
      num,
      hash,
      list
    );
  };
  $self->{dbh}->do($index);
}

# Overridden because Postgres uses SERIAL instead of AUTO_INCREMENT.
sub create_table_full_hashes_errors {
  my ($self, %args) = @_;

  my $schema = qq{
    CREATE TABLE full_hashes_errors (
      id SERIAL PRIMARY KEY,
      errors INT Default '0',
      prefix VARCHAR( 8 ),
      timestamp INT Default '0'
    );
  };

  $self->{dbh}->do($schema);
}


sub add_chunks_a {
  my ($self, %args) = @_;
  my $chunknum = $args{chunknum}  || 0;
  my $chunks   = $args{chunks}    || [];
  my $list     = $args{'list'}    || '';

  $self->{add_chunks_a_ins_sth} ||= $self->{dbh}->prepare('INSERT INTO a_chunks (hostkey, prefix, num, list) VALUES (?, ?, ?, ?)');
  $self->{add_chunks_a_del_sth} ||= $self->{dbh}->prepare('DELETE FROM a_chunks WHERE hostkey = ? AND  prefix  = ? AND num = ? AND  list  = ?');

  my $add = $self->{add_chunks_a_ins_sth};
  my $del = $self->{add_chunks_a_del_sth};

  foreach my $chunk (@$chunks) {
    # Crude workaround for longer prefixes. Although Google state that the
    # length varies, the overwhelming majority are 4 bytes. However, I have
    # seen a 32 byte one (chunk 69961).
    #
    # If this becomes more of a problem, the schema will have to be adjusted.
    if (length($chunk->{prefix}) > 8) {
      $chunk->{prefix} = substr $chunk->{prefix}, 0, 4;
    }

    $del->execute( $chunk->{host}, $chunk->{prefix}, $chunknum, $list );
    $add->execute( $chunk->{host}, $chunk->{prefix}, $chunknum, $list );
  }

  if (scalar @$chunks == 0) { # keep empty chunks
    $del->execute( '', '', $chunknum, $list );
    $add->execute( '', '', $chunknum, $list );
  }
}

=head1 SEE ALSO

See L<Net::Google::SafeBrowsing2> for handling Google Safe Browsing v2.

=head1 COPYRIGHT AND LICENSE

Copyright 2012 Nick Johnston, nickjohnstonsky@gmail.com. Based on
C<Net::Google::SafeBrowsing2> by Julien Sobrier.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=cut

1;


lib/Net/Google/SafeBrowsing2/Sqlite.pm
package Net::Google::SafeBrowsing2::Sqlite;

use strict;
use warnings;

use base 'Net::Google::SafeBrowsing2::DBI';

use Carp;
use DBI;
use List::Util qw(first);


our $VERSION = '0.9';


=head1 NAME

Net::Google::SafeBrowsing2::Sqlite - Sqlite as back-end storage for the Google Safe Browsing v2 database

=head1 SYNOPSIS

  use Net::Google::SafeBrowsing2::Sqlite;

  my $storage = Net::Google::SafeBrowsing2::Sqlite->new(file => 'google-v2.db');
  ...
  $storage->close();

=head1 DESCRIPTION

This is an implementation of L<Net::Google::SafeBrowsing2::Storage> using Sqlite.

=cut


=head1 CONSTRUCTOR

=over 4

=head2 new()

Create a Net::Google::SafeBrowsing2::Sqlite object

  my $storage = Net::Google::SafeBrowsing2::Sqlite->new(file => 'google-v2.db');

Arguments

=over 4

=item file

Required. File to store the database.

=item keep_all

Optional. Set to 1 to keep old information (such as expiring full hashes) in the database. 0 (delete) by default.

=item cache_size

Sqlite cache size. 20000 by default.

=back


=back

=cut

sub new {
	my ($class, %args) = @_;

	my $self = { # default arguments
		keep_all	=> 0,
		file		=> 'gsb2.db',
		cache_size 	=> 20000,

		%args,
	};

	bless $self, $class or croak "Can't bless $class: $!";


	$self->init();

    return $self;
}

=head1 PUBLIC FUNCTIONS

=over 4

See L<Net::Google::SafeBrowsing2::Storage> for a complete list of public functions.

=head2 close()

Cleanup old full hashes, and close the connection to the database.

  $storage->close();


=back

=cut

sub init {
	my ($self, %args) = @_;

	$self->{dbh} = DBI->connect("dbi:SQLite:dbname=" . $self->{file}, "", "");
	$self->{dbh}->do("PRAGMA journal_mode = OFF");
	$self->{dbh}->do("PRAGMA synchronous = OFF"); 
	$self->{dbh}->do("PRAGMA cache_size = " . $self->{cache_size}); 

	my @tables = $self->{dbh}->tables;

	if (! defined first { $_ eq '"main"."updates"' || $_ eq '"updates"' } @tables) {
		$self->create_table_updates();
	}
	if (! defined first { $_ eq '"main"."a_chunks"' ||  $_ eq '"a_chunks"' } @tables) {
		$self->create_table_a_chunks();
	}
	if (! defined first { $_ eq '"main"."s_chunks"' || $_ eq '"s_chunks"' } @tables) { 
		$self->create_table_s_chunks();
	}
	if (! defined first { $_ eq '"main"."full_hashes"' || $_ eq '"full_hashes"' } @tables) {
		$self->create_table_full_hashes();
	}
	if (! defined first { $_ eq '"main"."full_hashes_errors"' || $_ eq '"full_hashes_errors"' } @tables) { 
		$self->create_table_full_hashes_errors();
	}
	if (! defined first { $_ eq '"main"."mac_keys"' || $_ eq '"mac_keys"' } @tables) { 
		$self->create_table_mac_keys();
	}

	# Remove indexes created in previous versions
	$self->{dbh}->do('DROP INDEX IF EXISTS a_chunks_hostkey;');
	$self->{dbh}->do('DROP INDEX IF EXISTS a_chunks_num_list;');

	$self->{dbh}->do('DROP INDEX IF EXISTS s_chunks_hostkey;');
	$self->{dbh}->do('DROP INDEX IF EXISTS s_chunks_num;');
}


sub create_table_updates {
	my ($self, %args) = @_;

	my $schema = qq{	
		CREATE TABLE updates (
			last INTEGER DEFAULT 0,
			wait INTEGER DEFAULT 1800,
			errors INTEGER DEFAULT 0,
			list TEXT
		);
	}; # Need to handle errors

	$self->{dbh}->do($schema);
}

sub create_table_a_chunks {
	my ($self, %args) = @_;

	my $schema = qq{
		CREATE TABLE a_chunks (
			hostkey TEXT,
			prefix TEXT,
			num INTEGER,
			list TEXT
		);
	};

	$self->{dbh}->do($schema);

	my $index = qq{
		CREATE INDEX a_chunks_hostkey ON a_chunks (
			hostkey
		);
	};
# 	$self->{dbh}->do($index);

	$index = qq{
		CREATE INDEX a_chunks_num_list ON a_chunks (
			num,
			list
		);
	};
# 	$self->{dbh}->do($index);

	$index = qq{
		CREATE UNIQUE INDEX a_chunks_unique ON a_chunks (
			hostkey,
			prefix,
			num,
			list
		);
	};
	$self->{dbh}->do($index);
}

sub create_table_s_chunks {
	my ($self, %args) = @_;

	my $schema = qq{
		CREATE TABLE s_chunks (
			hostkey TEXT,
			prefix TEXT,
			num INTEGER,
			add_num INTEGER,
			list TEXT
		);
	};

	$self->{dbh}->do($schema);

	my $index = qq{
		CREATE INDEX s_chunks_hostkey ON s_chunks (
			hostkey
		);
	};
# 	$self->{dbh}->do($index);

	$index = qq{
		CREATE INDEX s_chunks_num ON s_chunks (
			num
		);
	};
# 	$self->{dbh}->do($index);

	$index = qq{
		CREATE UNIQUE INDEX s_chunks_unique ON s_chunks (
			hostkey,
			prefix,
			num,
			add_num,
			list
		);
	};
	$self->{dbh}->do($index);
}

sub create_table_full_hashes {
	my ($self, %args) = @_;

	my $schema = qq{
		CREATE TABLE full_hashes (
			id INTEGER PRIMARY KEY AUTOINCREMENT,
			num INTEGER,
			hash TEXT,
			list TEXT,
			timestamp INTEGER
		);
	};

	$self->{dbh}->do($schema);

	my $index = qq{
		CREATE UNIQUE INDEX hash ON full_hashes (
			num,
			hash,
			list
		);
	};
	$self->{dbh}->do($index);
}

sub create_table_full_hashes_errors {
	my ($self, %args) = @_;

	my $schema = qq{
		CREATE TABLE full_hashes_errors (
			id INTEGER PRIMARY KEY AUTOINCREMENT,
			errors INTEGER,
			prefix TEXT,
			timestamp INTEGER
		);
	};

	$self->{dbh}->do($schema);
}

sub create_table_mac_keys{
	my ($self, %args) = @_;

	my $schema = qq{
		CREATE TABLE mac_keys (
			client_key TEXT Default '',
			wrapped_key TEXT Default ''
		);
	};

	$self->{dbh}->do($schema);
}


sub add_chunks_s {
	my ($self, %args) 	= @_;
	my $chunknum		= $args{chunknum}	|| 0;
	my $chunks			= $args{chunks}		|| [];
	my $list			= $args{'list'}		|| '';

	my $add = $self->{dbh}->prepare('INSERT OR IGNORE INTO s_chunks (hostkey, prefix, num, add_num, list) VALUES (?, ?, ?, ?, ?)');

	$self->{dbh}->begin_work;
	foreach my $chunk (@$chunks) {
		$add->execute( $chunk->{host}, $chunk->{prefix}, $chunknum, $chunk->{add_chunknum}, $list );
	}

	if (scalar @$chunks == 0) { # keep empty chunks
		$add->execute( '', '', $chunknum, '', $list );
	}
	$self->{dbh}->commit;
}

sub add_chunks_a {
	my ($self, %args) 	= @_;
	my $chunknum		= $args{chunknum}	|| 0;
	my $chunks			= $args{chunks}		|| [];
	my $list			= $args{'list'}		|| '';

	my $add = $self->{dbh}->prepare('INSERT OR IGNORE INTO a_chunks (hostkey, prefix, num, list) VALUES (?, ?, ?, ?)');

	$self->{dbh}->begin_work;
	foreach my $chunk (@$chunks) {
		$add->execute( $chunk->{host}, $chunk->{prefix}, $chunknum, $list );
	}

	if (scalar @$chunks == 0) { # keep empty chunks
		$add->execute( '', '', $chunknum, $list );
	}
	$self->{dbh}->commit;
}


=head1 CHANGELOG

=over 4

=item 0.9

Keep empty sub chunks. Improve performances (remove sone idexes, use trnsactions).

New cache_size option.

=item 0.8

Index s_chunks_unique was created at the wrong place. Thanks to colinmkeith.

=item 0.7

Add option keep_all to keep expired full hashes. Useful for debugging.

=item 0.6

Use more efficient add_chunk_a and add_chunk_s functions.

=item 0.5

Use base class L<Net::Google::SafeBrowsing2::DBI>.

=item 0.4

Disable journalization. This speeds up updated by about 10x.

=item 0.3

Fix typos in the documentation.

=item 0.2

Add close() function to clean up old full hashes, and to close the connection to the database cleanly.

Add table and function to store and retrieve the Message Authentication Code (MAC) key.

In some environments, the module was trying to re-create existing tables. Fixed (Thank you to  Luis Alberto Perez).

=back

=head1 SEE ALSO

See L<Net::Google::SafeBrowsing2> for handling Google Safe Browsing v2.

See L<Net::Google::SafeBrowsing2::Storage> for the list of public functions.

See L<Net::Google::SafeBrowsing2::MySQL> for a back-end using Sqlite.

Google Safe Browsing v2 API: L<http://code.google.com/apis/safebrowsing/developers_guide_v2.html>


=head1 AUTHOR

Julien Sobrier, E<lt>jsobrier@zscaler.comE<gt> or E<lt>julien@sobrier.netE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Julien Sobrier

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.


=cut

1;